### DISCUSSION TITLE
关于Matcher中Rule与Permission执行先后顺序的问题

### DISCUSSION THREAD

**Discussion by ssttkkl** [nonebot/discussions#263] [想法]
目前发现Matcher为先执行check_perm再执行check_rule，但在我的理解里**检查权限通常比检查规则更为复杂，耗时更长**。因此是否先check_rule再check_perm更合理一点。

https://github.com/nonebot/nonebot2/blob/b25de93eb16c7a57199e85118846666578e0000a/nonebot/message.py#L128

比如为了权限控制，我需要将权限设置保存在文件或数据库中。此时先检查权限就会造成对于每一条消息的每一个Matcher都会去执行一遍取数据操作，这里造成的IO等待我感觉是无法容忍的。

**Comment by yanyongyu** [comment]
你可以使用[依赖注入缓存](https://v2.nonebot.dev/docs/next/advanced/di/dependency-injection#%E4%BE%9D%E8%B5%96%E7%BC%93%E5%AD%98)或者使用event preprocessor配合state进行存储来使得数据库操作仅执行一遍。在nb里权限先于规则，你可以理解为权限是你使用该应用的第一扇门，而规则是该应用对你不同的行为做出不同的响应。

例如：

```python
from nonebot.params import Depends

async def get_current_user(event: Event) -> User:
    ...  # only execute once

async def perm_1(user: User = Depends(get_current_user)) -> bool:
    # do something to check user

async def perm_2(user: User = Depends(get_current_user)) -> bool:
    # do something different
```
